benchexec/tools/jayhorn.py

```diff
@@ -0,0 +1,76 @@
+"""
+BenchExec is a framework for reliable benchmarking.
+This file is part of BenchExec.
+Copyright (C) 2007-2015  Dirk Beyer
+All rights reserved.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+import logging
+import xml.etree.ElementTree as ET
+import os
+import shutil
+import zipfile
+
+import benchexec.util as util
+import benchexec.tools.template
+import benchexec.result as result
+
+class Tool(benchexec.tools.template.BaseTool):
+    """
+    Tool info for JayHorn (https://github.com/jayhorn/jayhorn).
+    JayHorn does not support jar files yet;
+    we therefore unpack the benchmark jar file.
+    """
+
+    REQUIRED_PATHS = [
+                  "jayhorn.jar"
+                  "jayhorn" # contains: java -jar jayhorn.jar
+                  ]
+    def executable(self):
+        return util.find_executable('jayhorn')
+
+
+    def version(self, executable):
+        return "5.1" # --version not supported by JayHorn
+
+
+    def name(self):
+        return 'JayHorn'
+
+
+    def cmdline(self, executable, options, tasks, propertyfile, rlimits):
+        # unpack jar
+        os.mkdir('tmp')
```

Comment:
This approach has several problems:
- It clutters the user's working directory with files.
- It will cause huge problems with parallel benchmarks, either crashes or even wrong benchmark results without any warning.
- Removing `tmp` can delete files that the user put in this directory on purpose.

Why should input files be extracted before the benchmarking? If it makes sense to store these files in a compressed archive, why not let your tool extract them itself in memory?

Why should it be the responsibility of BenchExec to uncompress them, and not the user's or the tool's?

I saw the command above regarding JAR files, but it seems your tool is based on Java and reading ZIP files in Java is quite easy.

This will also lead to problems if JayHorn once learns how to uncompress JAR files, how would BenchExec know whether it should uncompress them or not?

In the end, of course it is your decision how you implement your tool, but the current proposal for the tool-info module creates these important problems for BenchExec users, so I would prefer to delegate the responsibility for uncompressing to the user if the tool can't handle it.

Reply:
I fully agree with everything said above. The wrapper just compensates for all the deficiencies of the tool itself. I'll have to sign up as a JayHorn and JPF contributor to resolve these issues.